### ISSUE TITLE
Named templates should be only accessible by the chart and its subcharts

### ISSUE THREAD

**Issue by luk355** [helm/helm#11085]
Consider a chart with 2 subcharts (_foo_, _bar_) as shown below. A named templates defined in _foo_ chart (i.e. `_foohelpers.tpl`) file should not be accessible from _bar_ chart. Allowing access to the named templates of another chart at the same level breaks the premise of ["A subchart is considered "stand-alone""](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/).

```
mychart/
├── charts/
│   ├── foo/
│   │   ├── templates/
│   │   │   ├── _foohelpers.tpl
│   │   │   └── foo.yaml
│   │   └── values.yaml
│   └── bar/
│       ├── templates/
│       │   ├── _barhelpers.tpl
│       │   └── bar.yaml
│       └── values.yaml
└── values.yaml
```

If _bar_ charts uses a named templated defined in _foo_ chart:

- `helm template mychart` will render all manifests just fine
- `helm template mychart/charts/bar` will fail as the referenced named template defined in _foo_ subchart will be unknown.

In case the same named template is defined in both _foo_ and _bar_ charts, one of them wins (operating system dependent I guess) and that one named template is used for both charts. This essentially leads to different manifest output of a _bar_ chart for `helm template mychart` and  `helm template mychart/charts/bar`.

Output of `helm version`:

version.BuildInfo{Version:"v3.9.0", GitCommit:"7ceeda6c585217a19a1131663d8cd1f7d641b2a7", GitTreeState:"clean", GoVersion:"go1.17.5"}

**Comment by bacongobbler** [comment]
Because chart dependencies are rendered in a flat-file structure and cannot be more than one level deep, we cannot support more sophisticated workflows such as this one.

To support this, we would need to support nested chart dependencies, and the chart dependency subsystem has not been worked on for a number of years since the core maintainer primarily responsible for that subsystem stepped down.

In other words, I'm in support of this change, but

1. It is a breaking change, and this cannot be implemented until Helm 4
2. It requires several major changes to the dependency subsystem for this to work as intended.
